app/lib/frontend/dom/dom.dart

```diff
@@ -135,17 +134,49 @@ Node codeSnippet({
 
 /// Creates a DOM element with ld+json `<script>` content.
 Node ldJson(Map<String, dynamic> content) {
-  final rawJson = json.encode(content);
-  final rawHtml = rawJson.replaceAllMapped(
-    _ldJsonEscapedCharactersRegExp,
-    (m) {
-      final code = m[0]!.codeUnitAt(0);
-      return r'\u' + code.toRadixString(16).padLeft(4, '0');
-    },
-  );
+  final sb = StringBuffer();
+
+  /// Build the JSON content by manually escaping most untrusted characters,
+  /// and also building the object and list structures.
+  void write(dynamic value) {
+    if (value is String) {
+      sb.write('"');
+      sb.write(value.replaceAllMapped(
+        _ldJsonEscapedCharactersRegExp,
+        (m) {
+          final code = m[0]!.codeUnitAt(0);
+          return r'\u' + code.toRadixString(16).padLeft(4, '0');
+        },
+      ));
+      sb.write('"');
+    } else if (value is List) {
+      sb.write('[');
+      for (var i = 0; i < value.length; i++) {
+        if (i > 0) sb.write(',');
+        write(value[i]);
+      }
+      sb.write(']');
+    } else if (value is Map) {
+      final entries = value.entries.toList();
+      sb.write('{');
+      for (var i = 0; i < entries.length; i++) {
+        if (i > 0) sb.write(',');
+        write(entries[i].key);
+        sb.write(':');
+        write(entries[i].value);
+      }
+      sb.write('}');
+    } else if (value is bool || value is num || value == null) {
+      sb.write(json.encode(value));
+    } else {
+      throw ArgumentError('Unexpected value: $value');
```

Comment:
Better explain that the `content` passed to `ldJson` could not be translated to JSON, because it contains a value of type `value.runtimeType` (or something like that)